### ISSUE TITLE
iOS 11 introduces smart quotes and smart dashes. May cause failure in call to enterText

### ISSUE THREAD

**Issue by mikelupo** [kif-framework/KIF#1012]
Testing on iOS versions prior to 11.0, this is fine,

However when typing into an input field in iOS 11.0,
```
[tester enterText:@"~!@#$%^&*()`1234567890-_+=[{]}:;'\"',<.>/?\\|" intoViewWithAccessibilityIdentifier:@"profile.FirstNameInputField"];
```

the test step fails due to the evaluation that takes place in KIFUITestActor.m
namely, this line fails.
```
KIFTestWaitCondition([actual isEqualToString:expected], error, @"Failed to get text \"%@\" in field; instead, it was \"%@\"", expected, actual);
```

When I look at the actual vs expected in the debugger I see a subtle difference. (namely, the characters just before ",<")
```
**expected:**  "~!@#$^&*()`1234567890-_+=[{]}:;'"',<.>/?\|"

**actual:**    "~!@#$^&*()`1234567890-_+=[{]}:;’”’,<.>/?\|"
```
@justinseanmartin @RoyalPineapple We can make an attempt to figure out what to do with the expected vs actual by doing something like this, 
```
    UITextView *textView = (UITextView *)view;
	if (@available(iOS 11.0, *)) {
		if (textView.smartQuotesType != UITextSmartQuotesTypeNo) //UITextSmartQuotesTypeDefault could pose another risk in that we'd probably have to make some assumptions.
		{
			//Do Something
		}

		if (textView.smartDashesType != UITextSmartDashesTypeNo) //UITextSmartDashesTypeNo could pose another risk in that we'd probably have to make some assumptions.
		{
			//Do Something
		}
	}
```
Discussions? Advice? Nitrous Oxide?

**Comment by justinseanmartin** [comment]
How would you feel about globally disabling `smartDashesType` and `smartQuotesType` on all text fields when running KIF tests? We've done the same internally for `autocorrectionType`, and were considering upstreaming that to avoid those issues that are outside of our control.

Right now, we do it via:
```
    // Prevent autocorrect in all UITextFields

    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        struct objc_method_description autocorrectionTypeMethodDescription = protocol_getMethodDescription(@protocol(UITextInputTraits), @selector(autocorrectionType), NO, YES);
        IMP noAutocorrectionTypeIMP = imp_implementationWithBlock(^(UITextField *_self) {
            return UITextAutocorrectionTypeNo;
        });
        class_replaceMethod([UITextField class], @selector(autocorrectionType), noAutocorrectionTypeIMP, autocorrectionTypeMethodDescription.types);
    });
```
(Thanks @harleyjcooper!)

I'm presuming that in general, these types of system behaviors aren't necessary or useful for verifying app behavior, because we can't control when/how it will kick in and it shouldn't alter normal app usage. 

That said, I could see cases where replacing a normal hyphen with an EM_DASH could actually break things like phone number formatting, so it might be bad to globally disable it.

A less invasive option for this particular case would be if the firstResponder field has smartDashes or smartQuotes enabled, we could have the `[KIFUITestActor expectView:toContainText:]` method use a regex to allow strings with mismatching smartQuotes/smartDashes to pass the check.

Thoughts?

**Comment by justinseanmartin** [comment]
I thought of one reason why simply patching `expectView:toContainText:` might not be good enough. 

If I'm entering form data one one screen and that gets auto-corrected to have smart punctuation, an action on another screen may also fail to find the text entered with what was assumed to be the entered non-smart punctuation.

**Comment by mikelupo** [comment]
I'm just manually turning it off in my own tests via the keyboard settings. 

Thinking out loud on the possibilities:

- Just turn it off for the sake of tests in the phone's settings, (i.e. General > Keyboard > Smart Punctuation = off) 
- A regex that's relaxed about it in the expectView:toContainText might be the right path.
- Perhaps get the trait on the field [myTextField smartQuotesType] and check for the returned enum, detect if we have the specific characters in the text entry and then take appropriate action based on the returned enum? (Not sure what assume in the default case)
[smartQuotesType](https://developer.apple.com/documentation/uikit/uitextsmartquotestype?language=objc)

**Comment by justinseanmartin** [comment]
Still doesn't deal with the scenario I called out where the test enters form data for something like device name being "Justin's iPhone" and it gets corrected to use smart quotes. A test that then wants to assert or interact with something that is driven by the entered form data will also need to be made aware that the text entered in the field may not be the same as the text entered by the test.

Do you have a programmatic solution to disabling the setting on the simulator? If so, do you do it from within the xctest environment or in CI prior to kicking off the tests?

Our solution so far has been to hard disable smart punctuation and autocorrect in the app via swizzling `UITextField` as outlined above.

**Comment by mikelupo** [comment]
@justinseanmartin  I used your method of globally disabling autocorrection (but customized for smartquotes and dashes) in my own test helper to disable the autocorrect feature. It did not have a positive affect on the test. (sorry, I don't have a working example of that code any more.)

Instead what I'm having to do is to first get the UITextField using a waitFor method. Then set the attributes on it before the keyboard appears. If I do it after the keyboard appears, then even that does not work.

Here is an example of what I do inside the test. This could be a work-around for anyone reading this post looking for such.

```
fnTextField = (UITextField *)[tester waitForViewWithAccessibilityIdentifier:@"FirstNameInputField"];
fnTextField.autocorrectionType = UITextAutocorrectionTypeNo;
	if (@available(iOS 11.0, *)) {
		fnTextField.smartDashesType = UITextSmartDashesTypeNo;
		fnTextField.smartQuotesType = UITextSmartQuotesTypeNo;
		fnTextField.spellCheckingType = UITextSpellCheckingTypeNo;
	}

//now enter a new first name.
	[tester enterText:@"~!@#$%^&*()`1234567890-_+=[{]}:;'\"',<.>/?\\|" intoViewWithAccessibilityIdentifier:@"profile.FirstNameInputField"];
```
